### PR TITLE
[BEAM-3419] Support iterable on Dataflow runner when using the unified worker.

### PR DESCRIPTION
**Pull request by lukecwik**
Note that all other portable runners are using iterable side inputs.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lukecwik** [comment]
Ran tests internal to Google to validate this change.

**Comment by lukecwik** [comment]
R: @tvalentyn @ananvay @robertwb

**Comment by ananvay** [comment]
Thanks Luke! LGTM

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 1]
```diff
@@ -320,24 +320,32 @@ def visit_transform(self, transform_node):
...
-              map_to_void_key = beam.pipeline.AppliedPTransform(
-                  pipeline,
-                  beam.Map(lambda x: (b'', x)),
-                  transform_node.full_label + '/MapToVoidKey%s' % ix,
-                  (side_input.pvalue,))
-              new_side_input.pvalue.producer = map_to_void_key
-              map_to_void_key.add_output(new_side_input.pvalue)
-              parent.add_part(map_to_void_key)
+              if use_unified_worker:
+                # Patch up the access pattern to appease Dataflow when using
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Please file a JIRA to get to the root of this discrepancy.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 2]
```diff
@@ -320,24 +320,32 @@ def visit_transform(self, transform_node):
...
+                # the UW and hardcode the output type to be Any since
+                # the Dataflow JSON and pipeline proto can differ in coders
+                # which leads to encoding/decoding issues within the runner.
+                side_input.pvalue.element_type = typehints.Any
+                new_side_input = _DataflowIterableSideInput(side_input)
```
Comment (robertwb) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Why is a special _DataflowIterableSideInput needed?

Comment (lukecwik) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: The dataflow translation logic relies on specific field names for the side input which differ from a Beam side input.

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py (hunk 1 above)
Comment (lukecwik) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Filed BEAM-9173

Comment (lukecwik) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py]: Added link in follow-up https://github.com/apache/beam/pull/10667

**Comment by tvalentyn** [comment]
LGTM

File: sdks/python/apache_beam/runners/dataflow/dataflow_runner.py [hunk 3]
```diff
@@ -320,24 +320,32 @@ def visit_transform(self, transform_node):
...
+              else:
+                # Add a map to ('', value) as Dataflow currently only handles
+                # keyed side inputs when using the JRH.
+                pipeline = side_input.pvalue.pipeline
+                new_side_input = _DataflowIterableAsMultimapSideInput(
+                    side_input)
```
Comment (TheNeuralBit) [inline sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:335]: Why does this switch to using `DataflowIterableAsMultimapSideInput` for the non-UW case? It looks like we used `DataflowIterableSideInput` here previously?

Comment (lukecwik) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:335]: _DataflowIterableSideInput was poorly named as it was using a multimap as the access pattern URN.

You can see in this change that _DataflowIterableSideInput was renamed to _DataflowIterableAsMultimapSideInput preserving the existing behavior.

Also the JRH (non-UW case) only supports the multimap access pattern.

Comment (TheNeuralBit) [reply sdks/python/apache_beam/runners/dataflow/dataflow_runner.py:335]: Got it, thank you!

**Pull request merged**